### PR TITLE
Add performance monitoring for diesel mechanic locomotives and new parameters

### PR DESCRIPTION
**Pull request by peternewell**
Improve the performance of diesel locomotives -

https://blueprints.launchpad.net/or/+spec/diesel-performance

This covers diesel mechanics and the following thread specifically will deal with any changes.

- http://www.elvastower.com/forums/index.php?/topic/35490-manual-gearbox-in-or/

### CODE REVIEW

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/GearBox.cs [hunk 1]
```diff
@@ -14,7 +15,11 @@ public class MSTSGearBoxParams
     {
         public int GearBoxNumberOfGears = 1;
         public int GearBoxDirectDriveGear = 1;
-        public int GearBoxType = 1;
+
+        // GearboxType ( 1 ) - power is continuous during gear changes (and throttle does not need to be adjusted) - this is the MSTS legacy operation so possibly needs to be the default.
+        // GearboxType ( 2 ) - power is interrupted during gear changes - but the throttle does not need to be adjusted when changing gear
+        // GearboxType ( 3 ) - power is interrupted and if GearboxOperation is Manual throttle must be closed when changing gear - see note on Wilson gearbox question below
+        public int GearBoxType = 1;
```
Comment (Sharpe49) [inline Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/GearBox.cs:22]: This variable should be an enumerated value.

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/GearBox.cs [hunk 2]
```diff
@@ -248,19 +253,41 @@ public void AutoAtGear()
...
+        /// <summary>
+        /// ClutchOn is true when clutch is fully engaged, and false when slipping
+        /// </summary>
         public bool clutchOn;
         public bool IsClutchOn
         {
             get
             {
-                if (DieselEngine.Locomotive.ThrottlePercent > 0)
+                if (Gears[0].TypeGearBox == 1)
```
Comment (Sharpe49) [inline Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/GearBox.cs:264]: In some cases, the variable is called TypeGearBox, in others, it is called GearBoxType.

This should be renamed GearBoxType

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/GearBox.cs [hunk 3]
```diff
@@ -14,7 +15,11 @@ public class MSTSGearBoxParams
     {
         public int GearBoxNumberOfGears = 1;
         public int GearBoxDirectDriveGear = 1;
-        public int GearBoxType = 1;
+
+        // GearboxType ( 1 ) - power is continuous during gear changes (and throttle does not need to be adjusted) - this is the MSTS legacy operation so possibly needs to be the default.
```
Comment (Sharpe49) [inline Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/GearBox.cs:19]: In programming, default should be 0.

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/GearBox.cs [hunk 4]
```diff
@@ -360,10 +390,26 @@ public float TractiveForceN
...
+                                // y = original amount ( 1 - decay rate)^length of prediction
+
+                                var decayRpM = ShaftRPM - DieselEngine.GovenorRPM;
+                                var teDecline = Math.Pow((1.0f - 0.05f), decayRpM);
+
+                                tractiveForceN = (float)Math.Abs(CurrentGear.MaxTractiveForceN * teDecline);
+                                tractiveForceN = MathHelper.Clamp(tractiveForceN, 0.0f, CurrentGear.MaxTractiveForceN);  // Clamp tractive effort so that it doesn't go below zero
+                            }
+
+//                            Trace.TraceInformation("Geared Tractive Effort #1 - Driving - TE: {0} lbf, RpM: {1}, Torque: {2} lb-ft, Throttle%: {3}, MaxTorque: {4} lb-ft, MaxTE {5} lbf, Speed: {6} mph, Clutch {7}, Gear: {8} IsClutchOn {9} Shaft RpM {10}", tractiveForceN * 0.224809f, DieselEngine.RealRPM, DieselEngine.DieselTorqueTab[DieselEngine.RealRPM] * 0.737562f, DieselEngine.DemandedThrottlePercent, DieselEngine.DieselTorqueTab.MaxY(), CurrentGear.MaxTractiveForceN * 0.224809f, CurrentSpeedMpS * 2.23694f, ClutchPercent, Locomotive.GearBoxController.CurrentNotch, IsClutchOn, ShaftRPM);
```
Comment (Sharpe49) [inline Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/GearBox.cs:411]: Deactivated debug code should not be in a comment, but in a preprocessor block.

For exemple:
#if GEARBOX_DEBUG
// Debug code
#endif

**Review by Sharpe49 — CHANGES_REQUESTED** [review]

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/GearBox.cs [hunk 5]
```diff
@@ -730,7 +729,7 @@ public void Update(float elapsedClockSeconds)
         }
     }
 
-    public enum TypeGearBox
+    public enum TypesGearBox
```
Comment (Sharpe49) [inline Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/GearBox.cs:732]: For this enum, you can also name it GearBoxTypeType or GearBoxTypeEnum.

Comment (peternewell) [reply Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/GearBox.cs:732]: Thanks for the suggestion.

The PR is still in draft, so I will respond to the other comments when I am happy to seek approval.

I have already taken onboard some of the suggestions.

Thanks

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerSupplies/DieselEngine.cs [hunk 6]
```diff
@@ -940,6 +953,24 @@ public void Initialize()
                 GearBox = new GearBox(this);
                 GearBox.Initialize();
             }
+
+            // It appears that the gearbox is initialised in two different places to cater for Basic and Advanced ENG file configurations(?).
+            // Hence the restore values recovered in gearbox class are being overwritten , and resume was not working
+            // Hence restore values are read as part of the diesel and restored at this point
```
Comment (Sharpe49) [inline Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerSupplies/DieselEngine.cs:973]: This is not how you should proceed:

In the current code, we save a boolean indicating if a gearbox has been saved. If this boolean is true, we create the GearBox object in the Restore function, assign the value, then we initialize.

```
bool gearSaved = inf.ReadBoolean();  // read boolean which indicates gear data was saved

if (gearSaved)
{
    GearBox = new GearBox(this);
    GearBox.Restore(inf);
}
```

You just have to modify lines 951 to 955 to check also is the GearBox is currently null and to Initialize if the GearBox object was created either here or in the Restore function.

**Review by Sharpe49 — CHANGES_REQUESTED** [review]

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerSupplies/DieselEngine.cs (hunk 6 above)
Comment (peternewell) [reply Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerSupplies/DieselEngine.cs:973]: > 
> You just have to modify lines 951 to 955 to check also is the GearBox is currently null and to Initialize if the GearBox object was created either here or in the Restore function.
>
Thanks for the feedback, and suggestion.

I am afraid that I don't quite understand your suggestion.

If you are able to provide me with a small patch file, I would be happy to incorporate it into my changes if it produces a correct restore.

**Comment by cjakeman** [comment]
Hi Peter,

The Code Bot https://james-ross.co.uk/projects/or reports a Merge conflict in Source/Orts.Simulation/Simulation/RollingStocks/MSTSDieselLocomotive.cs

Are you able to sort this out?

Thanks,

Chris

**Comment by peternewell** [comment]
> 
> Are you able to sort this out?
> 

Yes I was intending to delete and recreate the PR to overcome this issue.

Thanks

**Comment by Sharpe49** [comment]
You should use the rebase function of git to update your branch and then force push the branch to Github.

**Comment by peternewell** [comment]
> You should use the rebase function of git to update your branch and then force push the branch to Github.
>

Thanks for the suggestion.

For this one I prefer to rebuild it.

**Comment by peternewell** [comment]
PR closed to rebuild and overcome conflicts

**Pull request closed without merging**